documents/model.go

```diff
@@ -110,19 +123,353 @@ func (m *CoreDocumentModel) PrepareNewVersion(collaborators []string) (*CoreDocu
+	if err != nil {
+		return nil, err
+	}
+	// For every signature we create a LeafNode
+	sigProperty := proofs.NewProperty("signatures")
+	sigLeafList := make([]proofs.LeafNode, len(document.Signatures)+1)
+	sigLengthNode := proofs.LeafNode{
+		Property: sigProperty.LengthProp(),
+		Salt:     make([]byte, 32),
+		Value:    fmt.Sprintf("%d", len(document.Signatures)),
+	}
+	err = sigLengthNode.HashNode(h, false)
+	if err != nil {
+		return nil, err
+	}
+	sigLeafList[0] = sigLengthNode
+	for i, sig := range document.Signatures {
+		payload := sha256.Sum256(append(sig.EntityId, append(sig.PublicKey, sig.Signature...)...))
+		leaf := proofs.LeafNode{
+			Hash:     payload[:],
+			Hashed:   true,
+			Property: sigProperty.SliceElemProp(proofs.FieldNumForSliceLength(i)),
+		}
+		err = leaf.HashNode(h, false)
+		if err != nil {
+			return nil, err
+		}
+		sigLeafList[i+1] = leaf
+	}
+	err = tree.AddLeaves(sigLeafList)
+	if err != nil {
+		return nil, err
+	}
+	err = tree.Generate()
+	if err != nil {
+		return nil, err
+	}
+	return tree, nil
+}
+
+// CalculateDocumentRoot calculates the document root of the core document
+func (m *CoreDocumentModel) CalculateDocumentRoot() error {
+	document := m.Document
+	if len(document.SigningRoot) != 32 {
+		return errors.New("signing root invalid")
+	}
+
+	tree, err := m.GetDocumentRootTree()
+	if err != nil {
+		return err
+	}
+
+	document.DocumentRoot = tree.RootHash()
+	return nil
+}
+
+// getDataProofHashes returns the hashes needed to create a proof from DataRoot to SigningRoot. This method is used
+// to create field proofs
+func (m *CoreDocumentModel) getDataProofHashes(dataRoot []byte) (hashes [][]byte, err error) {
+	tree, err := m.GetDocumentSigningTree(dataRoot)
+	if err != nil {
+		return
+	}
+
+	signingProof, err := tree.CreateProof("data_root")
+	if err != nil {
+		return
+	}
+
+	rootProofHashes, err := m.getSigningRootProofHashes()
+	if err != nil {
+		return
+	}
+
+	return append(signingProof.SortedHashes, rootProofHashes...), err
+}
+
+// getSigningRootProofHashes returns the hashes needed to create a proof for fields from SigningRoot to DocumentRoot. This method is used
+// to create field proofs
+func (m *CoreDocumentModel) getSigningRootProofHashes() (hashes [][]byte, err error) {
+	tree, err := m.GetDocumentRootTree()
+	if err != nil {
+		return
+	}
+	rootProof, err := tree.CreateProof("signing_root")
+	if err != nil {
+		return
+	}
+	return rootProof.SortedHashes, err
+}
+
+// CalculateSigningRoot calculates the signing root of the core document
+func (m *CoreDocumentModel) CalculateSigningRoot(dataRoot []byte) error {
+	doc := m.Document
+	tree, err := m.GetDocumentSigningTree(dataRoot)
+	if err != nil {
+		return err
+	}
+
+	doc.SigningRoot = tree.RootHash()
+	return nil
+}
+
+// AccountCanRead validate if the core document can be read by the account .
+// Returns an error if not.
+func (m *CoreDocumentModel) AccountCanRead(account identity.CentID) bool {
```

Comment:
Can you add `NFTOwnerCanRead` as well ?

Reply:
I am not entirely clear how you imagined the NFT interactions with the CoreDocModel to be, in particular the token registry and how CoreDocuments are currently used in the ethereum payment obligation files. This is why I haven't moved these functionalities yet.